src/control/server/server.go

```diff
@@ -77,6 +77,7 @@ type server struct {
 	grpcServer   *grpc.Server
 
 	onEnginesStarted []func(context.Context) error
+	onShutdown       []func()
```

Comment:
This is a valid consideration, although I could see an argument being made that there isn't really any meaningful error handling to do on shutdown, and the main reason for supplying a context to the other callbacks is to allow for context-based shutdown of any goroutines started by them. In this case any passed-in context would likely already be canceled and these callbacks shouldn't be starting any new goroutines or otherwise doing blocking stuff that might need to select on `ctx.Done()`.

Reply:
@tanabarr I actually started out with the `func(context.Context) error` signature, and pulled back for exactly the reasons Mike described. I would be okay with re-adding the error to the signature if you prefer, and having the server.shutdown() call log it. That was the only thing I could think of to do with errors.